add-controllers

Chegulov opened the pull request:
добавлены CRUD методы для User
добавлены GET, POST, PATCH для Item
добавлены исключения и их обработчик.

Chegulov committed:
add-controllers
добавлены CRUD методы для User
добавлены GET, POST, PATCH для Item
добавлены исключения и их обработчик.

Chegulov committed:
add-controllers
исправлен codeStyle

TyutterinYakov commented on src/main/java/ru/practicum/shareit/Error/ErrorHandler.java:
```diff
@@ -0,0 +1,63 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public  ErrorResponse handleWrongAccesException(final WrongAccesException e) {
```
Так же можно было бы не создавать для каждого исключения, обрабатываемого как `NOT_FOUND`, отдельные хендлеры, а собрать их в одну группу в `@ExceptionHandler({WrongAccesException.class, DataNotFoundException.class...})`
А принимаемым аргументом метода-обработчика будет общий класс-родитель - `RuntimeException`.

Chegulov replied:
Сделано.

TyutterinYakov commented on src/main/java/ru/practicum/shareit/Error/ErrorHandler.java:
```diff
@@ -0,0 +1,63 @@
+
+@Slf4j
+@RestControllerAdvice
+public class ErrorHandler {
```
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем `Throwable`, а возвращать статус 500 INTERNAL_SERVER_ERROR

Chegulov replied:
Сделано.

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,46 @@
+    }
+
+    @GetMapping("/{id}")
+    public ItemDto getItemById(@RequestHeader("X-Sharer-User-Id") int userId, @PathVariable int id) {
```
Название заголовка хорошо было бы вынести в отдельный класс в виде статической константы, чтобы в случае изменений - произвести их в одном месте) Кстати, статические константы записываются заглавными буквами, а слова разделяются нижней чертой
https://www.oracle.com/java/technologies/javase/codeconventions-contents.html

Chegulov replied:
Сделано.

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,46 @@
+import java.util.List;
 
 /**
  * TODO Sprint add-controllers.
```
Если функционал класса реализован - комментарий следует исключить, либо же следует описать то, что остается реализовать) Аналогично в других местах)

Chegulov replied:
Сделано.

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,46 @@
 /**
  * TODO Sprint add-controllers.
  */
+@Slf4j
```
Аннотация есть, а логгирование отсутствует

Chegulov replied:
Убрал аннотацию

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemService.java:
```diff
@@ -0,0 +1,46 @@
+
+    public List<ItemDto> getItemByText(int userId, String text) {
+        userStorage.getUsersById(userId);
+        if (text == null || text.isBlank()) {
```
Проверка на `null` не имеет смысла, так как параметр запроса является обязательным, так как в `@RequestParam` не указано обратное - `defaultValue = ".."` или `required = false`

Chegulov replied:
Убрал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemService.java:
```diff
@@ -0,0 +1,46 @@
+    public List<ItemDto> getItemByText(int userId, String text) {
+        userStorage.getUsersById(userId);
+        if (text == null || text.isBlank()) {
+            return new ArrayList<>();
```
Кстати, вместо создания нового объекта класса `ArrayList`, можно было бы вернуть пустую неизменяемую коллекцию полученную из `List.of()` или `Collections.empty()`

Chegulov replied:
Сделал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,46 @@
+    private final ItemService itemService;
+
+    @PostMapping
+    public ItemDto create(@RequestHeader("X-Sharer-User-Id") int userId, @Valid @RequestBody Item item) {
```
Для представления тела запроса - следует также использовать не классы-сущности, а классы-DTO, так как мы не должны нарушать изоляцию) Классы-модели не должны использовать на транспортном уровне

Chegulov replied:
Исправил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,47 @@
+    }
+
+    @PostMapping
+    public UserDto create(@Valid @RequestBody User user) {
```
Для представления тела запроса - следует также использовать не классы-сущности, а классы-DTO, так как мы не должны нарушать изоляцию) Классы-модели не должны использовать на транспортном уровне

Chegulov replied:
Исправил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,47 @@
 /**
  * TODO Sprint add-controllers.
  */
+@Slf4j
```
Аннотация присутствует, но нет логгирования :(

Chegulov replied:
Убрал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,47 @@
+    }
+
+    @PostMapping
+    public UserDto create(@Valid @RequestBody User user) {
```
Кстати, вместо `@Valid` можно использовать такую аннотацию `@Validated(Create.class)`. Таким образом мы указываем, какую группу аннотаций следует учитывать. `Create` - это интерфейс, который ты создаешь самостоятельно, он будет пустым и служит лишь для указания группы) Аналогично следует поступить при обновлении, но там будет аннотация `@Validated(Update.class)`

Если не использовать валидацию по группам, тогда потребуется при обновлении дополнительно проверять формат переданной строки с почтой в коде, что не очень удобно. Либо же потребуется ввести отдельный класс DTO для обновления, чтобы над полем с почтой использовать `@Email`

Chegulov replied:
Добавил, спасибо

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,15 @@
+
+@Data
+@Builder
+public class UserDto {
```
Так же и здесь, можно было бы использовать валидацию по группам)
Для имени:
`@NotBlank(groups = {Create.class})`
Для почты аннотацию `@Email` следует учитываться для обновления и создания, так как, если она не будет передана вообще, то все будет окей, а если будет, тогда она будет проверена на соответствие почте, а вот аннотацию
`@NotEmpty` только для создания) Так как при обновлении передавать ее не обязательно)

Chegulov replied:
Добавил, спасибо

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,22 @@
+
+@Data
+@Builder
+public class User {
```
Аннотации валидации более не требуются, так как теперь они все будут находиться в DTO)

Chegulov replied:
Убрал

TyutterinYakov replied:
Аннотации валидации более не требуются, так как теперь они все будут находиться в DTO)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,24 @@
+
+@Data
+@Builder
 public class Item {
```
Аннотации валидации более не требуются, так как теперь они все будут находиться в DTO, так как именно он будет использоваться на транспортном уровне)

Chegulov replied:
Убрал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/storage/InMemoryItemStorage.java:
```diff
@@ -0,0 +1,65 @@
+
+@RequiredArgsConstructor
+@Component
+public class InMemoryItemStorage implements ItemStorage {
```
(предложение) Для упрощения работы можно ввести еще одну мапу такого вида:
`private final Map<Long, List<Item>> userItemIndex = new LinkedHashMap<>();`
В ней ключами будут выступать идентификаторы пользователей, а в виде значений, будут их вещи) Таким образом, если нам потребуется получить все вещи пользователя, нам достаточно будет обратиться к этой мапе, без итерации по всем значениям другой)

При сохранении можно будет использовать такую запись)
`final List<Item> items = userItemIndex.computeIfAbsent(item.getOwner().getId(), k -> new ArrayList<>());`
Таким образом, благодаря методу computeIfAbsent будет создан новый объект ArrayList'a, если его нет в мапе по ключу, либо же вернется ссылка на него, если он будет найден) А далее мы уже сможем в него добавить новый предмет пользователя)

Chegulov replied:
Добавил `private final Map<Integer, Map<Integer, Item>> userItemIndex;` для быстрого поиска и обновления.

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/storage/InMemoryItemStorage.java:
```diff
@@ -0,0 +1,65 @@
+@Component
+public class InMemoryItemStorage implements ItemStorage {
+    private final Map<Integer, Item> items;
+    private final ItemMapper itemMapper;
```
Уровень доступа к данным должен возвращать объекты класса-сущности(Item), а не DTO, так как мы не должны нарушать изоляцию) 
https://ru.stackoverflow.com/questions/1110180/Почему-репозиторий-не-должен-возвращать-dto

Chegulov replied:
Перенёс маппер в сервис

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,77 @@
+public class InMemoryUserStorage implements UserStorage {
+
+    private final Map<Integer, User> users;
+    private final UserMapper userMapper;
```
Уровень доступа к данным должен возвращать объекты класса-сущности(Item), а не DTO, так как мы не должны нарушать изоляцию) 
https://ru.stackoverflow.com/questions/1110180/Почему-репозиторий-не-должен-возвращать-dto

Chegulov replied:
Перенёс маппер в сервис

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/storage/InMemoryItemStorage.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public ItemDto getItemById(int userId, int id) {
```
(Предложение) Хорошо было бы использовать в методах получения по идентификатору тип `Optional` https://habr.com/ru/post/346782/ Гарантирую, ты не пожалеешь, а заодно в будущем тебе будет проще, так как он активно используется при работе со Spring Data JPA

Chegulov replied:
Изменил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/storage/InMemoryItemStorage.java:
```diff
@@ -0,0 +1,65 @@
+
+        Item item = items.get(id);
+        item = itemMapper.updateItemFromDto(item, itemDto);
+        items.put(id, item);
```
Вызов метода `put` не требуется, так как мы работаем с ссылкой на объект, а значит все изменения будут зафиксированы в самом объекте и обновление ссылки на ту, что ссылается на этот же объект - ничего не изменит)

Chegulov replied:
вынес маппер в сервис

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public UserDto getUsersById(int id) {
```
(Предложение) Хорошо было бы использовать в методах получения по идентификатору тип `Optional` https://habr.com/ru/post/346782/ Гарантирую, ты не пожалеешь, а заодно в будущем тебе будет проще, так как он активно используется при работе со Spring Data JPA

Chegulov replied:
Изменил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,77 @@
+            throw new DuplicateDataException("Пользователь с таким email уже зарегистрирован");
+        }
+        user = userMapper.updateUserFromDto(user, userDto);
+        users.put(id, user);
```
Вызов метода `put` не требуется, так как мы работаем с ссылкой на объект, а значит все изменения будут зафиксированы в самом объекте и обновление ссылки на ту, что ссылается на этот же объект - ничего не изменит)

Chegulov replied:
В данном случае put нужен. Так как для проверки на дублирование email я удаляю пользователя
`User user = users.get(id);
        users.remove(id);`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/storage/InMemoryItemStorage.java:
```diff
@@ -0,0 +1,65 @@
+    public List<ItemDto> getItemByText(String text) {
+        String finalText = text.toLowerCase();
+        return items.values().stream()
+                .filter(item -> (item.getDescription().toLowerCase().contains(finalText)
```
(доп информация) Описание условий в одном `filter` более эффективно)
https://www.baeldung.com/java-streams-multiple-filters-vs-condition

Chegulov replied:
Исправил

TyutterinYakov reviewed (commented):
Доброе утро! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше! Хорошего дня :) Если возникнут вопросы - пиши в пачку, буду рад помочь🌅

Chegulov committed:
add-controllers
перенесёны мапперы в сервисы
добавлен Optional<T> для getById методов
внесены мелкие исправления

Chegulov committed:
add-controllers
исправлен codyStyle

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/storage/InMemoryItemStorage.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public List<Item> getItems(int userId) {
+        return new ArrayList<>(userItemIndex.get(userId).values());
```
Вместо метода `get` лучше использовать `getOrDefault`, чтобы избежать проблем с NPE, если у пользователя отсутствуют вещи и никогда их не было)
`return new ArrayList<>(userItemIndex.getOrDefault(userId, Map.of()).values());`

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного вечера🌃🎇

Chegulov committed:
add-controllers
убраны аннотации валидации в User
